AzureTable/Microsoft.DataTransfer.AzureTable/Sink/TableAPIBulkSinkAdapter.cs

```diff
@@ -0,0 +1,172 @@
+
+            inputSizeTracker.Add(item);
+
+            if (!dict.ContainsKey(item.PartitionKey))
+            {
+                dict[item.PartitionKey] = new TableBatchOperation();
+            }
+
+            var batchOperation = dict[item.PartitionKey];
+            batchOperation.InsertOrReplace(item);
+
+            if (inputSizeTracker.HasExceededInputLimits())
+            {
+                await FlushToTableAsync(cancellation);
+            }
+        }
+
+        public async Task CompleteAsync(CancellationToken cancellation)
+        {
+            if (!inputSizeTracker.HasExceededInputLimits())
+            {
+                await FlushToTableAsync(cancellation);
+            }
+        }
+
+        public void Dispose()
+        {
+        }
+
+        // Flush out all data to the Cosmos DB Table sink.
+        private async Task FlushToTableAsync(CancellationToken cancellation)
+        {
+            try
+            {
+                List<Exception> exceptions = new List<Exception>();
+
+                foreach (var kv in dict)
+                {
+                    var batchOperation = kv.Value;
+                    if (batchOperation.Count > 0)
+                    {
+                        var subOperations = batchSizeTracker.Split(batchOperation);
+
+                        foreach (var subOperation in subOperations)
+                        {
+                            var op = new TableBatchOperation();
+                            for (int i = 0; i < subOperation.Count(); i++)
+                            {
+                                op.Add(subOperation.ElementAt(i));
+                            }
+
+                            try
+                            {
+                                await Utils.ExecuteWithRetryAsync(
+                                                    () => cloudtable.ExecuteBatchAsync(op, cancellation)
+                                                );
+                            }
+                            catch (Exception ex)
+                            {
+                                string listofDocumentsNotCommitted = string.Join(",", op.Select(x => x.Entity.RowKey));
+                                ex =  new Exception(
+                                        string.Format("{0} : offending documents having PartitionKey={1}: RowKeys:[{2}]",
+                                        ex.Message, op[0].Entity.PartitionKey, listofDocumentsNotCommitted), ex
+                                );
+                                exceptions.Add(ex);
+                            }
+                        }
+                    }
+                }
+
+                if (exceptions.Count > 0)
+                {
+                    throw new AggregateException(exceptions);
+                }
+            }
+            finally
+            {
+                dict.Clear();
+                batchSizeTracker.Clear();
```

Comment:
Why not clear the inputSizeTracker directly here?

Reply:
inputSizeTracker and BatchSizeTracker go hand-in-hand.
The relationship is that when the input exceeds the configured input-buffering limit, we will flush out the accumulated records in batches.
When the batch is complete batch-tracker is done, and it takes up the responsibility of clearing up the associated inputtracker -- which in a way, codifies the relationship.
We could clear it here too -- it does not matter.

Reply:
Spoke with Kiran. No need to fix.